File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
n = int(input())
numbers = list(map(int, input().split()))
nums = [0]
start, end, mid = 0,0,0
for num in numbers:
    if nums[-1]<num:
        nums.append(num)
    else:
        start = 0
        end = len(nums)
        while start<end:
            mid = (start+end)//2
            if nums[mid]<num:
                start=mid+1
            else:
                end = mid
        nums[end]=num
print(len(nums)-1)

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()
	var n, start, end, mid int
	fmt.Fscanf(reader, "%d\n", &n)
	var numbers = make([]int, n)
	var nums = []int{0}
	for i := 0; i < n; i++ {
		fmt.Fscanf(reader, "%d ", &numbers[i])

	}
	for _, v := range numbers {
		if nums[len(nums)-1] < v {
			nums = append(nums, v)
		} else {
			start = 0
			end = len(nums)
			for start < end {
				mid = int((start + end) / 2)
				if nums[mid] < v {
					start = mid + 1
				} else {
					end = mid
				}
			}
			nums[end] = v
		}
	}
	fmt.Fprintf(writer, "%d", len(nums) - 1)
}
